packages/coil-extension/src/background/services/Stream.ts

```diff
@@ -211,14 +222,86 @@ export class Stream extends EventEmitter {
     return details
   }
 
-  async _stream() {
-    this._lastOutgoingMs = Date.now()
+  onMoney(data: {
+    sentAmount: string
+    amount: number
+    assetCode: string
+    assetScale: number
+    sourceAmount: string
+    sourceAssetCode: string
+    sourceAssetScale: number
+  }) {
+    if (data.amount <= 0) return
+
+    this._debug(`emitting money. amount=${data.amount}`)
+    const now = Date.now()
+    const msSinceLastPacket = now - this._lastOutgoingMs
+    this._lastOutgoingMs = now
+    const event: StreamMoneyEvent = Object.assign(data, {
+      paymentPointer: this._paymentPointer,
+      packetNumber: this._packetNumber++,
+      requestId: this._requestId,
+      initiatingUrl: this._initiatingUrl,
+      msSinceLastPacket: msSinceLastPacket,
+      amount: data.amount.toString()
+    })
+    this.emit('money', event)
+  }
+
+  async stop() {
+    this._active = false
+    if (this._attempt) {
+      await this._attempt.stop()
+      this._attempt = null
+    }
+  }
+
+  async pause() {
+    this.stop()
+  }
+
+  async resume() {
+    this.start()
+  }
+
+  private async abort() {
+    // Don't call this.stop() directly, let BackgroundScript orchestrate the
+    // stop.
+    this.emit('abort', this._requestId)
+  }
+}
+
+interface StreamAttemptOptions {
+  bandwidth: AdaptiveBandwidth
+  onMoney: (event: any) => void
+  requestId: string
+}
+
+let ATTEMPT = 0
 
-    const [plugin, spspDetails] = await Promise.all([
-      this._getPlugin(),
-      this._getSPSPDetails()
-    ])
+class StreamAttempt {
```

Comment:
Nice. It's always kind of bugged me that the Stream class represented multiple streams and the way it handled state. Will look at this closer today.